### PR TITLE
Made example input file which measures interface energy

### PR DESCRIPTION
**Pull request by jbair34**
Closes issue #5867

### CODE REVIEW

**Comment by dschwen** [comment]
please check your commit again

**Comment by jbair34** [comment]
Check it for what? it is squashed into one commit now according to my log...

**Comment by permcody** [comment]
It still contains a libMesh update. We can't merge that. You'll need to soft reset your commit and remove your libMesh update.

**Comment by jbair34** [comment]
@permcody So I should git reset HEAD^ and then re-commit it and push it?

**Comment by permcody** [comment]
Almost...

After you reset HEAD^, you'll want to run `git submodule update`. This will checkout the existing submodule hash. You can verify by doing another `git status` where you should only see your other file sitting there which you can then recommit and push.

**Comment by jbair34** [comment]
Thanks @permcody  I think that commit should be better... I didn't see anything out of place in the git status...

**Comment by dschwen** [comment]
It would be nice if we could figure out a way to make the size of the cell easily configurable. Right now it appears in 
* the initial condition
* the Int_energy MOOSE function 
* one PointValue postprocessor

One suggestion would be to use ```SideAverageValue``` instead of ```Pointvalue```. this would permit you to specify boundary names, instead of the domain size dependent point.

**Comment by dschwen** [comment]
Ok, to recap: Use an ```ElementIntegralMaterialProperty``` with ```mat_prop = 1``` to calculate the length of your simulation cell and write a ramp IC that creates a linear ramp from a min to max value across the simulation cell (query the x extent of the mesh for this).

**Comment by jbair34** [comment]
So I am working on the IC but quick question... ElementIntegralMaterialProperty is again a post processor so won't it not work just like the numelems post processor didn't work?

**Comment by dschwen** [comment]
But we will not be using that PP in the IC! It is for use in the function that computes the interfacial free energy.

**Comment by jbair34** [comment]
Right... got it.

**Comment by jbair34** [comment]
Is there an example somewhere for how to query the x extent of the mesh? I haven't found anything in the other ICs like that...

**Comment by dschwen** [comment]
```
MooseMesh::dimensionWidth(unsigned int component)
```
is what you need. Search through the code to find out how to access the mesh object.

**Comment by permcody** [comment]
@jbair34 - you should probably squash these down again. At least the commits that represent bug fixes to your PR.

File: modules/phase_field/examples/measure_interface_energy/1Dinterface_energy.i [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    f_name = F
+    function = c^2*(c-1)^2
+    args = c
+    derivative_order = 2
+    block = 0
+  [../]
+[]
+
+[Postprocessors]
+  # The total free enrgy of the simulation cell to observe the energy reduction.
```
Comment (dschwen) [inline modules/phase_field/examples/measure_interface_energy/1Dinterface_energy.i]: typo: enrgy

File: modules/phase_field/include/ics/RampIC.h [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+  /**
+   * The value of the gradient at a point.
+   */
+  virtual RealGradient gradient(const Point &p);
+
+  Real _xlength;
+  Real _xmin;
+  Real _value_left;
+  Real _value_right;
```
Comment (dschwen) [inline modules/phase_field/include/ics/RampIC.h]: those members can all be ```const```

File: modules/phase_field/examples/measure_interface_energy/1Dinterface_energy.i [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    type = ElementIntegralVariablePostprocessor
+    variable = c
+  [../]
+  # Get simulation cell size (1D volume) from postprocessor
+  [./volume]
+    type = ElementIntegralMaterialProperty
+    block = 0
+    mat_prop = 1
+  [../]
+  # Find concentration in each phase using max and min
```
Comment (dschwen) [inline modules/phase_field/examples/measure_interface_energy/1Dinterface_energy.i]: No, this is _wrong_! We do not know a priori if the concnetrations are the min and max, or the max and min respectively! You need to be consistent with the Fleft and Fright PPs! Use SideAverageValue.

Comment (jwpeterson) [inline modules/phase_field/examples/measure_interface_energy/1Dinterface_energy.i]: > No, this is wrong!

I think you mean "Nein!"

**Comment by dschwen** [comment]
Please add a test for the RampIC as well (preferably with an odd simulation cell xmin = -5 xmax = 13)

File: modules/phase_field/examples/measure_interface_energy/1Dinterface_energy.i (hunk 3 above)
Comment (dschwen) [inline modules/phase_field/examples/measure_interface_energy/1Dinterface_energy.i]: 9
9
9
9
!!

**Comment by dschwen** [comment]
We test ICs in the phase field module with ```solve = false```. Take a look at some of the tests there. There is no need to have any kernels and any solve in these tests.

**Comment by jbair34** [comment]
I updated that IC but just so you know in phase_field/tests/CH_IC there are a whole bunch of IC tests that are Transient with a couple of time steps running... Do you want me to change all of those to solve = false?

**Comment by dschwen** [comment]
No, we'll do a separate issue for those.

**Comment by dschwen** [comment]
I think @permcody fixed the option to disable the kernel coverage check. If so you can remove all kernels (add ```kernel_coverage_check = false``` in the ```[Problem]``` block). You can remove the ```w``` variable and its kernels anyways.

**Comment by permcody** [comment]
@dschwen - another review please!

File: modules/phase_field/tests/RampIC/RampIC_test.i [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  [../]
+  [./Free_energy]
+    type = DerivativeParsedMaterial
+    f_name = F
+    function = c^2*(c-1)^2
+    args = c
+    derivative_order = 2
+    block = 0
+  [../]
+[]
```
Comment (dschwen) [inline modules/phase_field/tests/RampIC/RampIC_test.i]: I don't think we need a Materials block here.

File: modules/phase_field/src/base/PhaseFieldApp.C [hunk 5]
```diff
@@ -293,6 +294,7 @@ PhaseFieldApp::registerObjects(Factory & factory)
   registerInitialCondition(SpecifiedSmoothCircleIC);
   registerInitialCondition(ThumbIC);
   registerInitialCondition(Tricrystal2CircleGrainsIC);
+  registerInitialCondition(RampIC);
```
Comment (dschwen) [inline modules/phase_field/src/base/PhaseFieldApp.C]: The include and register statements are sorted alphabetically. Let's keep it that way.

File: modules/phase_field/include/ics/RampIC.h [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+class RampIC : public InitialCondition
+{
+public:
+  RampIC(const InputParameters & parameters);
+
+protected:
+  /**
+   * Evaluate the function at the current quadrature point and timestep.
+   */
+  const Real f();
```
Comment (dschwen) [inline modules/phase_field/include/ics/RampIC.h]: What is this? It does not appear in the implementation file.

File: modules/phase_field/include/ics/RampIC.h [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+//Forward Declarations
+class RampIC;
+class MooseMesh;
+
+template<>
+InputParameters validParams<RampIC>();
+
+/**
+ * Defines a boundary condition that forces the value to be a user specified
```
Comment (dschwen) [inline modules/phase_field/include/ics/RampIC.h]: No, it does not. Please make sure that copy pasted comments are updated to reflect what the code actually does.

File: modules/phase_field/include/ics/RampIC.h [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+
+#ifndef RAMPIC_H
+#define RAMPIC_H
+
+#include "InitialCondition.h"
+#include "InputParameters.h"
+#include "FEProblem.h"
+#include "MooseMesh.h"
```
Comment (dschwen) [inline modules/phase_field/include/ics/RampIC.h]: neither MooseMesh.h nor FEProblem.h need to be included in the header. We should avoid unnecessary include statements as they will increase compile time.

File: modules/phase_field/tests/RampIC/RampIC_test.i [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+  type = FEProblem
+  solve = false
+  kernel_coverage_check = false
+[]
+
+[Executioner]
+  type = Steady
+  solve_type = PJFNK
+  petsc_options_iname = '-pc_type -pc_hypre_type'
+  petsc_options_value = 'hypre boomeramg'
```
Comment (dschwen) [inline modules/phase_field/tests/RampIC/RampIC_test.i]: if we do not solve we won't need PETSc options...

File: modules/phase_field/tests/RampIC/tests [hunk 10]
```diff
@@ -0,0 +1,8 @@
+[Tests]
+  [./RampIC_test]
```
Comment (dschwen) [inline modules/phase_field/tests/RampIC/tests]: I'm not a fan of tests that are called ```*test```. But let's just leave it that way. I plan on reorganizing the entire phase_field module tests directory anyways... if I end up with too much time on my hands.

**Pull request merged**